Enabling tracing by usage of opentelemetry

rem1-dev opened the pull request:
I've added tracing to the project by introducing mainly telemetry, opentelemetry and Tokyo crates. Having tracing allows analysis of bottleneck functions (functions that take majority of execution time).

shesek committed:
Optimize and refactor lookup_txos()

- Avoid unnecessary copying of prev outpoints

- When looking for both mempool and on-chain txos, accumulate the set of
  outpoints that remain to be looked up to avoid re-checking for them
  later again in the found set

- Refactored lookup_txos() to use lookup_txo() internally rather than
  the other way around, which was less efficient

- Lookup txos in mempool first, then on-chain

- ChainQuery::lookup_txos() now returns a Result instead of panicking
  when outpoints are missing

- Removed ChainQuery::lookup_avail_txos() and allow_missing, which are
  no longer neceesary

YusukeShimizu committed:
nix: add binLiquid for flake.nix

add binLiquid to the inherit list to enable specific builds

rikublock committed:
add support for blockchain.scripthash.unsubscribe

shesek committed:
Don't use RPC batching with bitcoind

This actually hurts performance because the batched response has to be
bueffered on the bitcoind side, as @TheBlueMatt explains at
https://github.com/romanz/electrs/issues/373#issuecomment-785533444

Instead, send multiple individual RPC requests in parallel using a
thread pool, with a separate RPC TCP connection for each thread.

Also see https://github.com/romanz/electrs/pull/374

shesek committed:
Continuously attempt to fetch mempool transactions

Previously, if any of the mempool transactions were not available
because they were evicted between getting the mempool txids and txs
themselves, the mempool syncing operation would be aborted and tried
again from scratch.

With this change, we instead keep whatever transactions we were able to
fetch, then get the updated list of mempool txids and re-try fetching
missing ones continuously until we're able to get a full snapshot.

shesek committed:
Don't store mempool txs before all prevouts are available

The indexing process was adding transactions into the store so that
prevouts funded & spent within the same batch could be looked up via
Mempool::lookup_txos(). If the indexing process later failed for any
reason, these transactions would remain in the store.

With this change, we instead explicitly look for prevouts funded within
the same batch, then look for the rest in the chain/mempool indexes and
fail if any are missing, without keeping the transactions in the store.

shesek committed:
Make sure the chain tip doesn't move while fetching the mempool

shesek committed:
Update logging verbosity

- Reduce logging level for bitcoind's JSONRPC response errors

  These can happen pretty often for missing mempool txs and do not
  warrant warn-level logging. Unexpected RPC errors will bubble up and
  be handled appropriately.

- Add more verbose logging for mempool syncing

shesek committed:
Reuse RPC threads and TCP connections

Keep RPC TCP connections open between sync runs and reuse them,
to reduce TCP connection initialization overhead.

shesek committed:
Avoid recomputing txids when possible

Following https://github.com/Blockstream/electrs/pull/89#discussion_r1632848608
and https://github.com/Blockstream/electrs/pull/89#discussion_r1632919266

RCasatta committed:
upgrade electrumd dep

RCasatta committed:
upgrade electrum, do tests in nix

Now on nix env we provide the env vars of the executables needed for
integration testing, so we can enable tests.

To be coherent with the electrum nix version used, upgrade also the
autodownloaded one.

Note we have to change a test assertion, it seems electrum behaviour
changed, upgrading balances before confirmation.

RCasatta commented:
Isn't `prometheus` trying to achieve the same thing? Should we remove that if we decide to use opentelemetry?

RCasatta committed:
Avoid print logs in tests

to instead have logs like before use `RUST_LOG=debug cargo test`

RCasatta committed:
introduce benchmarking

Using criterion so that we don't need nightly to run benchmark.

The "bench" features is also introduced as a way to bench private
methods via public methods feature gated via this feature.

RCasatta committed:
Avoid recomputing txids

on my machine:

add_blocks              time:   [4.7562 ms 4.7754 ms 4.7972 ms]
                        change: [-15.059% -14.650% -14.169%] (p = 0.00 < 0.05)
                        Performance has improved.

RCasatta committed:
Avoid recomputing txid in TxConfRow

add_blocks              time:   [3.9458 ms 3.9586 ms 3.9717 ms]
                        change: [-17.564% -17.103% -16.660%] (p = 0.00 < 0.05)
                        Performance has improved.

RCasatta committed:
Avoid recomputing txid in TxRow

add_blocks              time:   [2.9788 ms 2.9938 ms 3.0101 ms]
                        change: [-24.810% -24.373% -23.897%] (p = 0.00 < 0.05)
                        Performance has improved.

RCasatta committed:
print test logs in CI

philippem commented:
> Isn't `prometheus` trying to achieve the same thing? Should we remove that if we decide to use opentelemetry?

- prometheus gives us metrics ( example there is a histogram of times spent making RPC calls to bitcoind). Many alerts can be directly built on prometheus metrics.
- opentelemtry (OTLP) gives us tracing within and between processes (for example we can look at a mempool update operation and see where the time is being spent)

They both enhance observability but are complementary.

shesek commented:
> > Isn't `prometheus` trying to achieve the same thing? Should we remove that if we decide to use opentelemetry?
> 
> ...
> They both enhance observability but are complementary.

I'm still not quite sure I understand the differences. Prometheus can and is being used to track function execution times, for example in these functions (via `_timer` and its `Drop`) where opentelemetry was added too:

```diff
+    #[instrument(skip_all, name="Mempool::utxo")]
     pub fn utxo(&self, scripthash: &[u8]) -> Vec<Utxo> {
         let _timer = self.latency.with_label_values(&["utxo"]).start_timer();
         let entries = match self.history.get(scripthash) {
@@ -209,6 +216,7 @@ impl Mempool {
             .collect()
     }
 
+    #[instrument(skip_all, name="Mempool::stats")]
     // @XXX avoid code duplication with ChainQuery::stats()?
     pub fn stats(&self, scripthash: &[u8]) -> ScriptStats {
         let _timer = self.latency.with_label_values(&["stats"]).start_timer();
@@ -258,12 +266,14 @@ impl Mempool {
         stats
     }
 
+    #[instrument(skip_all, name="Mempool::txids")]
     // Get all txids in the mempool
     pub fn txids(&self) -> Vec<&Txid> {
         let _timer = self.latency.with_label_values(&["txids"]).start_timer();
         self.txstore.keys().collect()
     }
```

shesek commented on src/bin/electrs.rs:
```diff
@@ -16,7 +16,7 @@ use electrs::{
     metrics::Metrics,
     new_index::{precache, ChainQuery, FetchFrom, Indexer, Mempool, Query, Store},
-    rest,
+    otlp_trace, rest,
```
Missing newline (should probably `cargo fmt` the other changes too)

shesek commented on src/electrum/server.rs:
```diff
@@ -623,7 +643,7 @@ impl Connection {
     pub fn run(mut self, receiver: Receiver<Message>) {
         self.stats.clients.inc();
-        conditionally_log_rpc_event!(self, json!({ "event": "connection established" }));
+        //conditionally_log_rpc_event!(self, json!({ "event": "connection established" }));
```
Was this intentionally disabled?

rem1-dev replied:
I reverted it

shesek commented on src/config.rs:
```diff
@@ -378,7 +378,7 @@ impl Config {
         } else {
             stderrlog::Timestamp::Off
         });
-        log.init().expect("logging initialization failed");
```
Why?

rem1-dev replied:
At some point of searching for working tracing implementation this was needed to remove runtime error. I think something else was already initialising log. Anyway in current version this is not a problem any more. I restored this line.

shesek commented on Cargo.toml:
```diff
@@ -51,7 +51,13 @@ url = "2.2.0"
 # close to same tokio version as dependent by hyper v0.14 and hyperlocal 0.8 -- things can go awry if they mismatch
-tokio = { version = "1", features = ["sync", "macros"] }
+tokio = { version = "1", features = ["sync", "macros", "rt-multi-thread", "rt"] }
+opentelemetry = { version = "0.20.0", features = ["rt-tokio"] }
```
6 new (direct) dependencies, oh my :sweat_smile:

Should we perhaps put this behind a feature flag so that the extra dependencies could be avoided by users that don't require tracing?

rem1-dev replied:
I added a feature flag to this. By default tracing is off and can be turned on with:
`--no-default-features --features otlp-tracing`

shesek commented:
It's not critical, but it could aid review if the addition of `#[instrument()]` attributes was separated to its own commit

RCasatta committed:
Flag to enable compaction during initial sync

While compaction during initial sync makes thing much slower, it may be
preferred to not require to size the disk the double of the final
required space.

mario-rei committed:
Introducing opentelemetry tracing dependencies

mario-rei committed:
Opentelemetry initialization function impl

RCasatta committed:
improve logging during initial sync

it is less interesting to see how many rows are written in the db and
more interesting knowing the last height indexed

mario-rei committed:
Instrumenting functions

mario-rei committed:
Make telemetry optional

philippem commented on src/daemon.rs:
```diff
@@ -16,6 +16,8 @@ use serde_json::{from_str, from_value, Value};
 use bitcoin::consensus::encode::{deserialize, serialize_hex};
 #[cfg(feature = "liquid")]
 use elements::encode::{deserialize, serialize_hex};
+#[cfg(feature = "tracing-enabled")]
```
the feature name should maybe be "otlp-tracing" or "tracing". "enabled" is what we do with it.

rem1-dev replied:
I changed the name

philippem commented on Cargo.toml:
```diff
@@ -14,6 +14,14 @@ edition = "2018"
 [features]
 liquid = [ "elements" ]
 electrum-discovery = [ "electrum-client"]
+tracing-enabled = [
```
let's call this "tracing"

rem1-dev replied:
That makes feature name a cyclic dependency with tracing crate. Therefore feature name will be:
```
otlp-tracing = [
    "tracing",
```

It could be workaround by specifying "dep:"
```
tracing = [
    "dep:tracing",
```
but that syntax  on the other hand doesn't work with "tracing/max_level_trace", so this would fail:
```
tracing = [
    "dep:tracing",
    "dep:tracing/max_level_trace",
```

shesek commented:
How do you feel about defining a no-op `#[instrument]` macro when the tracing feature is disabled, so that we don't have to wrap it in `#[cfg_attr(feature = "tracing",... ))` every time?

RCasatta committed:
Use `electrs` as default run

By doing so simply `cargo run` launch the server, while now it is
required `cargo run --bin electrs`

mario-rei committed:
Renaming flag for turning on tracing

RCasatta committed:
getblocks retry 5 times on 'Block not found on disk'

mario-rei committed:
Make tracing macro no-op by default

mario-rei committed:
Restoring log init

rem1-dev commented:
> How do you feel about defining a no-op `#[instrument]` macro when the tracing feature is disabled, so that we don't have to wrap it in `#[cfg_attr(feature = "tracing",... ))` every time?

I found out that instead of defining my own procedural macro I can just use `tracing/max_level_off` to turn off tracing by default and make `#[instrument]` macro no-op by default. I implemented it.

rem1-dev commented:
Tracing is turned off by default and can be turned on by providing following params for both `cargo build` and `cargo run` commands:

```
--no-default-features --features otlp-tracing
```

shesek commented:
It would be nice if it was possible to avoid the additional dependencies altogether, not just disable tracing

rem1-dev commented:
> It would be nice if it was possible to avoid the additional dependencies altogether, not just disable tracing

I guess that needs, as you suggested, a new macro that is no-op in case of otlp-tracing feature flag is off. You're right that we still have dependency on new crates. But by default the only new crates we will depend on is just one: tracing - so that we can turn off it's macro that we used. Only if we turn on future flag we pull other dependencies in addition to tracing:
"tracing-subscriber",
"opentelemetry",
"tracing-opentelemetry",
"opentelemetry-otlp",
"opentelemetry-semantic-conventions"

so we add 1 new dependency only (or 6 if one decides to use tracing). Is it still not acceptable @shesek ?

shesek commented:
I had a preference to avoid additional dependencies altogether, and thought this should be as simple as

```rust
#[not(cfg(feature = "otlp"))]
#[proc_macro_attribute]
pub fn instrument(_attr: TokenStream, item: TokenStream) -> TokenStream { item }
```

However I realized that this would require making this a `proc-macro` crate would would prevent us from exporting anything other than proc macros, or make a separate crate just for this, which overly complicates things... So yeah, let's stick with the one extra dependency.

shesek commented on Cargo.toml:
```diff
@@ -13,9 +13,23 @@ edition = "2018"
 bench = []
+default = ["no-otlp-tracing"]
+otlp-tracing = [
+    "tracing",
```
The "tracing" crate can be made non-optional instead of enabling it in both cases.

rem1-dev replied:
removed

mario-rei committed:
removing redundand dependency

shesek commented on src/daemon.rs:
```diff
@@ -203,6 +211,8 @@ impl Connection {
         })
     }
 
+
```
nit: extra newline

shesek commented on src/electrum/server.rs:
```diff
@@ -13,6 +13,9 @@ use error_chain::ChainedError;
 use hex::{self, DisplayHex};
 use serde_json::{from_str, Value};
 
+
```
nit: extra newline

shesek commented on src/electrum/server.rs:
```diff
@@ -601,6 +620,8 @@ impl Connection {
         }
     }
 
+
```
nit: extra newline

shesek commented on src/new_index/query.rs:
```diff
@@ -11,6 +11,9 @@ use crate::errors::*;
 use crate::new_index::{ChainQuery, Mempool, ScriptStats, SpendingInput, Utxo};
 use crate::util::{is_spendable, BlockId, Bytes, TransactionStatus};
 
+
```
nit: extra newline

shesek commented on src/new_index/schema.rs:
```diff
@@ -375,9 +384,8 @@ impl ChainQuery {
 
+    #[instrument(skip_all, name="sdchema::Indexer::get_block_txids")]
     pub fn get_block_txids(&self, hash: &BlockHash) -> Option<Vec<Txid>> {
-        let _timer = self.start_timer("get_block_txids");
```
Was this removed intentionally?

rem1-dev replied:
Probably it's manual merge conflict resolution error. I'll restore it.

shesek commented on src/new_index/schema.rs:
```diff
@@ -797,14 +828,18 @@ impl ChainQuery {
             .map(BlockId::from)
     }
 
+
```
nit: extra newline

shesek commented on src/new_index/schema.rs:
```diff
@@ -995,6 +1045,7 @@ fn add_blocks(block_entries: &[BlockEntry], iconfig: &IndexerConfig) -> Vec<DBRo
         .collect()
 }
 
+
```
nit: extra newline

shesek commented on src/new_index/schema.rs:
```diff
@@ -1070,7 +1124,9 @@ fn index_blocks(
         .collect()
 }
 
+
```
nit: extra newline

shesek commented on src/rest.rs:
```diff
@@ -24,6 +24,9 @@ use tokio::sync::oneshot;
 use std::fs;
 use std::str::FromStr;
 
+
```
nit: extra newline

shesek commented on src/bin/electrs.rs:
```diff
@@ -142,6 +145,25 @@ fn run_server(config: Arc<Config>) -> Result<()> {
+    #[cfg(feature = "otlp-tracing")]
+    let _tracing_guard = otlp_trace::init_tracing("electrs");
+
+    let config = Arc::new(Config::from_args());
```
To avoid repeating this code here and in `main()` and to avoid `allow(dead_code)`, I would restructure it like this:

```rust
fn main_() {
    let config = Arc::new(Config::from_args());
    // ... rest of body
}

#[cfg(feature = "no-otlp-tracing")]
fn main() { main_() }

#[cfg(feature = "otlp-tracing")]
#[tokio::main]
async fn main() {
    let _tracing_guard = otlp_trace::init_tracing("electrs");
    main_()
}
```

rem1-dev replied:
Good point!

shesek reviewed (changes requested):
Added a suggestion and some nits (with apologies for the newline pedantry 😅)

It's kind of unfortunate that we have to use a negative feature, which requires using `--no-default-features` to disable `no-otlp-tracing` when enabling `otlp-tracing`. It'll become more annoying if we ever add other default features, as they'll have to be re-enabled manually. But I don't see a way around it... is there? (/cc @RCasatta)

shesek commented:
> It's kind of unfortunate that we have to use a negative feature ...

One thing we could do to make this less of a footgun is to add some explicit useful error messages when `--no-default-features` is used without re-enabling either otlp features, or when `otlp-tracing` is enabled without using `--no-default-features`:

```rust
#[cfg(not(any(feature = "otlp-tracing", feature = "no-otlp-tracing")))]
compile_error!("Must enable one of the 'otlp-tracing' or 'no-otlp-tracing' features");
#[cfg(all(feature = "otlp-tracing", feature = "no-otlp-tracing"))]
compile_error!("Cannot enable both the 'otlp-tracing' and 'no-otlp-tracing' features");
```

RCasatta commented:
> It's kind of unfortunate that we have to use a negative feature, which requires using --no-default-features to disable no-otlp-tracing when enabling otlp-tracing. It'll become more annoying if we ever add other default features, as they'll have to be re-enabled manually. But I don't see a way around it... is there? (/cc @RCasatta)

By looking here https://github.com/tokio-rs/tracing/blob/bdbaf8007364ed2a766cca851d63de31b7c47e72/tracing/src/level_filters.rs#L68 it seems that maximum tracing is the default so we don't need to opt-in for it? Thus, it seems to me we can avoid the positive feaure? 
It would require to use code like `not(feature = "no_otlp_tracing")` which doesn't look great but at least we would need only 1 feature and continue to guarantee that features are additive.

shesek commented:
> Thus, it seems to me we can avoid the positive feature?

The original goal for using the feature was to make (most) of the additional dependencies optional, which (I think?) we can't do with just a negative feature

RCasatta commented:
I see... I understand the issue of having too many deps, but I think it's greater in "money-handling" libs such as wallets than in electrs, so I would bite the bullet and build some deps even though they are not used instead of adding complexity.

rem1-dev commented:
> > Thus, it seems to me we can avoid the positive feature?
> 
> The original goal for using the feature was to make (most) of the additional dependencies optional, which (I think?) we can't do with just a negative feature

Yes, that was the motivation behind this, keep additional dependencies at minimum and also make #instrument[] annotations no-op by default. I couldn't find other solution than `--no-default-features --features otlp-tracing` and turning off tracing in default features.

shesek commented on src/bin/electrs.rs:
```diff
@@ -153,8 +153,6 @@ fn main_() {
 }
 
-#[cfg(not(any(feature = "otlp-tracing", feature = "no-otlp-tracing")))]
-compile_error!("Must enable one of the 'otlp-tracing' or 'no-otlp-tracing' features");
```
Why was this check removed? If we're keeping the features it would be better to cover both failure cases.

Also, I would place this in `lib.rs`.

rem1-dev replied:
I'll move it. Regarding that removed condition, I just couldn't trigger it. Always the second compile check was triggered

shesek replied:
It should be triggered if you use `--no-default-features` without enabling anything, doesn't it?

rem1-dev replied:
You're right it is. I missed it because it also causes a flood of 
"error: cannot determine resolution for the attribute macro `instrument`" errors as expected. I added this check too.

shesek commented:
>  the issue of having too many deps, but I think it's greater in "money-handling" libs such as wallets than in electrs

I generally agree with the sentiment, but its a lot of new dependencies - 6 direct ones, expanded into a total of 71 new crates in the dependency graph - for a feature that won't typically be used by other users of esplora/electrs.

Adding just the `tracing` crate as implemented by @rem1-dev reduces it to just 2 new dependencies (`tracing`+ `tracing-core`).

shesek commented:
I miscalculated the crates in my previous comment - the full set of dependencies expands into 41 new crates in the graph, while `no-otlp-tracing`doesn't add any new ones at all (`tracing` is already a sub-dependency via `hyper`).

rem1-dev commented:
@shesek github summary for this PR shows there is still  a "requested changes" but I can't find anything. Is there anything else I should change?

shesek commented:
> github summary for this PR shows there is still a "requested changes" but I can't find anything

I don't see anything either, seems that all the requests were resolved :+1:

Could you run `cargo fmt` and perhaps squash some of the commits? It's ready to go other than that.

rem1-dev commented:
> > github summary for this PR shows there is still a "requested changes" but I can't find anything
> 
> I don't see anything either, seems that all the requests were resolved 👍
> 
> Could you run `cargo fmt` and perhaps squash some of the commits? It's ready to go other than that.

I run `cargo fmt`. Unfortunately squashing would be a bigger effort because  when I try to rebase and squash commits there are lots of merge conflicts. I don't understand that since I already resolved conflicts when merging new-index into my branch. Anyway can we skip squashing?

shesek commented:
@rem1-dev Yes it's fine, I can just squash to a single commit when merging.

However it seems that there are merge conflicts, can you look into that?

rem1-dev commented:
> @rem1-dev Yes it's fine, I can just squash to a single commit when merging.
> 
> However it seems that there are merge conflicts, can you look into that?

Are there? Github says "This branch has no conflicts with the base branch". When I tried merge new-index to my branch locally also git said all is up to date. I did small change: amended last commit message.

RCasatta commented:
I confirm I cannot find conflicts, @shesek I'll let you merge if you want to squash

rem1-dev commented:
Let's not merge it yet. I'll try again and solve merge conflict during commits squash.

rem1-dev commented:
I squashed commits. PR must be approved again.

rem1-dev commented:
Something wen't awfully wrong during that commits squash and new-index branch merge. Not mine commits are now in commits list. I think best to do now is if I create fresh branch from `new-index` and apply my changes there. Let's close this PR and I'll create new one. Sorry for that.

shesek commented:
--> #128
